processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java

```diff
@@ -477,6 +478,87 @@ public Row apply(Object input)
           }
         };
       }
+
+      @Override
+      public Function<Row, Object> prepareForResultLevelCache()
+      {
+        return new Function<Row, Object>()
+        {
+          @Override
+          public Object apply(Row input)
+          {
+            if (input instanceof MapBasedRow) {
+              final MapBasedRow row = (MapBasedRow) input;
+              final List<Object> retVal = Lists.newArrayListWithCapacity(1 + dims.size() + aggs.size());
+              retVal.add(row.getTimestamp().getMillis());
+              Map<String, Object> event = row.getEvent();
+              for (DimensionSpec dim : dims) {
+                retVal.add(event.get(dim.getOutputName()));
+              }
+              for (AggregatorFactory agg : aggs) {
+                retVal.add(event.get(agg.getName()));
+              }
```

Comment:
can we refactor code a little bit so that above code is not repeated in both here and `prepareForCache()` ?
same comment for other places this is done.

Reply:
also it would nice to have UTs for these methods.

Reply:
UTs added. Attempted to refactor but it resulted in increasing the complexity of prepare and pullFromResultLevelCache. In multiple cases, the Result row has to be cycled through in both pullFromCache and pullFromResultLevelCache in order to deserialize the post agg data. We might have to rewrite prepareforcache as well to do a proper refactor, which I thought wasn't a good idea at this point.